test_qmk_compiler.py

```diff
@@ -0,0 +1,337 @@
+fixture here is a non-trivial task and will take work that the author has
+not had time to put in yet.
+"""
+import filecmp
+import os
+import os.path
+import re
+import shutil
+from tempfile import mkstemp
+
+import pytest
+
+import update_kb_redis
+import qmk_commands
+
+
+############################################################################
+# Setup Environment                                                        #
+############################################################################
+
+
+def test_0000_checkout_qmk_skip_cache():
+    """Make sure that we successfully git clone qmk_firmware and generate the version.txt hash.
+    """
+    qmk_commands.checkout_qmk(skip_cache=True)
+    assert os.path.exists('qmk_firmware/version.txt')
+
+
+############################################################################
+# Begin Tests                                                              #
+############################################################################
+
+
+# Source code retrieval testing. Make sure we're storing and fetching the
+# correct source every way we can. At least test_0001 must be run before any
+# other test in this file or they will all fail.
+def test_0001_fetch_source_qmk():
+    """Make sure that the qmk_firmware.zip the previous test uploaded is the same as the one on disk.
+    """
+    os.rename('qmk_firmware.zip', 'qmk_firmware_cloned.zip')
+    qmk_commands.fetch_source('qmk_firmware')
+    assert filecmp.cmp('qmk_firmware_cloned.zip', 'qmk_firmware.zip')
+    os.remove('qmk_firmware_cloned.zip')
+    os.remove('qmk_firmware.zip')
+
+
+def test_0002_checkout_qmk_cache():
+    """Make sure that we fetch the cache from QMK and don't clone from git.
+    """
+    git_hash = open('qmk_firmware/version.txt', 'r').read()
+    shutil.rmtree('qmk_firmware')
+    update_kb_redis.checkout_qmk(require_cache=True)
+    cached_hash = open('qmk_firmware/version.txt', 'r').read()
+    assert git_hash == cached_hash
+
+
+def test_0003_checkout_qmk_skip_and_require_cache():
+    """Make sure that we can't pass skip_cache and require_cache at the same time
+    """
+    with pytest.raises(ValueError):
+        update_kb_redis.checkout_qmk(skip_cache=True, require_cache=True)
+
+
+# Test out functions that require qmk_firmware be in place on disk.
+def test_0010_keyboard_list():
+    """Test qmk_commands.keyboard.list()
+    """
+    keyboard_list = update_kb_redis.list_keyboards()
+    assert len(keyboard_list) > 0
+
+    for keyboard in keyboard_list:
+        assert os.path.exists(os.path.join('qmk_firmware', 'keyboards', keyboard))
+
+
+def test_0011_find_firmware_file_hex():
+    """Make sure that qmk_commands.find_firmware_file() can find a hex file.
+    """
+    fd, test_firmware = mkstemp(suffix='.hex', dir='.')
+    firmware_file = qmk_commands.find_firmware_file()
+    os.remove(test_firmware)
+    assert os.path.split(test_firmware)[-1] == firmware_file
```

Comment:
probably a dumb question, but what generates the hex file?

Reply:
`mkstemp()` does.